### PR TITLE
Topic modelling

### LINKED ISSUE #71: Optional: Topic Modeling
Participants are always very keen on topic modelling since it is relatively easy to see the results without much data preparation. Introducing topic modeling at the end of episode 2 might be the most natural place, since it also uses gensim. BERTopic also is an option but this would be after BERT is introduced.

### PR DESCRIPTION
**Pull request by carschno**
Add a sub-section about topic modelling.

### CODE REVIEW

File: episodes/02-word_representations.md [hunk 1]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+For the sake of modelling, however, a topic is typically represented as a weighted list of _words_ like this:
+
+| Topic 1 | Topic 2 | Topic ... |
+| ------- | ------- | --- |
+| cat     | boat    | ... |
+| dog     | car     | ... |
+| ...     | ...     | ... |
+
+A text, on the other hand, is represented as a weighted list of _topics_, for instance:
```
Comment (angel-daza) [inline episodes/02-word_representations.md]:
```suggestion
A document, on the other hand, is represented as a weighted list of _topics_, for instance:
```

File: episodes/02-word_representations.md [hunk 2]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+| Document 1 | Document 2 | Document ... |
+| ---------- | ---------- | ------------ |
+| Topic 5    | Topic 8    | ... |
+| Topic 1    | Topic 3    | ... |
+| Topic 2    | Topic 5    | ... |
+
+The term `document` is equally under-specified in this context.
```
Comment (angel-daza) [inline episodes/02-word_representations.md]:
```suggestion
The term `document` is also under-specified in this context, but usually refers to a chunk of text, its size depending on the convenience of granularity (sentence, paragraph, book chapter, etc..)
```

File: episodes/02-word_representations.md [hunk 3]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+::: callout
+
+### Unsupervised Learning and Self-Supervised Learning
+
+In machine learning terminology, topic modelling algorithms apply _unsupervised learning_:
+they do not rely on any human annotations about the data that could be used to _supervise_ the modelling process.
+Instead, topic modelling forms _clusters_ of data points by grouping similar documents together.
+
+Algorithms widely differ in how they measure similarity and how they form clusters.
```
Comment (angel-daza) [inline episodes/02-word_representations.md]:
```suggestion
Clustering algorithms widely differ in how they measure similarity and how they form clusters.
```

File: episodes/02-word_representations.md [hunk 4]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+The term `document` is equally under-specified in this context.
+A topic model typically uses sentences or paragraphs to approximate topics.
+This is partly due to computational constraints: processing longer documents requires significantly more processing power and memory.
+
+Another benefit of smaller text units come from linguistic insights: topics tend to be most consistent on the level of sentences or paragraphs, while they vary throughout a longer text.
+While topic models do allow for texts to convey multiple topics, a model learns best if the topics are separated from each other as clearly as possible.
+
+::: callout
+
+### Unsupervised Learning and Self-Supervised Learning
```
Comment (angel-daza) [inline episodes/02-word_representations.md]: not very clear to me from this callout what "self-supervised" learning is

File: episodes/02-word_representations.md [hunk 5]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+Algorithms widely differ in how they measure similarity and how they form clusters.
+Regardless, topic modelling algorithms have been applied in domains unrelated to linguistics, like [biological](https://link.springer.com/article/10.1007/s10661-025-14291-0) and [climate research](https://cnrs.hal.science/hal-04484617v1/file/Article_CMIP6_LDA-2.pdf).
+
+:::
+
+### Model Training
+
+There are various implementations of topic modelling.
+Before modern language models were available, models based on word co-occurrences were dominant, most prominently as LDA (_Latent Dirichlet Allocation_).
```
Comment (angel-daza) [inline episodes/02-word_representations.md]:
```suggestion
Before modern language models were available, models based on word co-occurrences were dominant, most prominently LDA (_Latent Dirichlet Allocation_).
```

File: episodes/02-word_representations.md [hunk 6]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+It represents both topics and documents as probability distributions across words and topics respectively, based on word frequencies and distributions in the training data.
```
Comment (angel-daza) [inline episodes/02-word_representations.md]:
```suggestion
In LDA each topic is represented as a probability distribution over words and conversely, each document has a probability distribution of topics, based on observed words and sampling from distributions across the full corpus.
```

File: episodes/02-word_representations.md [hunk 7]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+::: callout
+
+### Topic Model Implementations
+
+Topic modelling has been a relevant tool for text analysis for decades.
+Before modern (large) language models were available, algorithms like LSI (_Latent Semantic Indexing_) and LDA (_Latent Dirichlet Allocation_) have applied different statistical and probabilistic methods on word counts to generate word, document and topic representations from any document collection.
+The Gensim library provides implementations for both [LSI](https://radimrehurek.com/gensim/models/lsimodel.html) and [LDA](https://radimrehurek.com/gensim/models/ldamodel.html).
+Various flavours of these methods have implemented methods to consider, for instance, metadata or time sequences into the models.
+
+The newer generation of topic modelling uses language models like BERT to generate contextual word representations.
```
Comment (angel-daza) [inline episodes/02-word_representations.md]:
```suggestion
The newer generation of topic modelling uses contextualized language models like BERT to generate contextual word representations.
```

File: episodes/02-word_representations.md [hunk 8]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+```python
+from bertopic import BERTopic
+
+# TODO: determine good number based on full data
+N_SENTENCES = 100000
+
+topic_model = BERTopic(nr_topics=100)
+topics, probs = topic_model.fit_transform(sentences[:N_SENTENCES])
+
+Loading weights: 100%|██████████| 103/103 [00:00<00:00, 7405.86it/s]
```
Comment (angel-daza) [inline episodes/02-word_representations.md:657]: Is this "logging info" here on purpose?

File: episodes/02-word_representations.md [hunk 9]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+The `nr_topics=100` argument above is not strictly needed.
+We use it in this example to keep the output manageable, because without the default settings result in approximately 900 different topics on this dataset.
+
+Output the topics that the algorithm has identified:
+
+```python
+topic_model.get_topic_info()
+
+	Topic	Count	Name	Representation	Representative_Docs
```
Comment (angel-daza) [inline episodes/02-word_representations.md:677]: Perhaps this is the same as above... but shouldn't this part be inside its own output block? If it stays like this it might be confusing if they copy paste it into their notebook

File: episodes/02-word_representations.md [hunk 10]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+4	3	1625	3_katharine_carrie_cassandra_said	[katharine, carrie, cassandra, said, william, ...	[think katharine, katharine, katharine]
+...	...	...	...	...	...
+95	94	16	94_isabella_prison_quadroon_sold	[isabella, prison, quadroon, sold, previous, u...	[immediately took daughter aside informed prev...
+96	95	12	95_foreman_workshop_ogre_motioning	[foreman, workshop, ogre, motioning, gait, bar...	[head foreman round stopped machine, foreman, ...
+97	96	12	96_ames_superiority_funniest_thorne	[ames, superiority, funniest, thorne, dora, sa...	[going asked ames, said ames, ames]
+98	97	11	97_clatter_bang_scatter_crash	[clatter, bang, scatter, crash, lads, dogs, fi...	[clatter clatter clatter bang, clatter clatter...
+99	98	11	98_entanglement_ensue_faithlessness_repugnant	[entanglement, ensue, faithlessness, repugnant...	[entanglement carrie anxious change sort, carr...
+```
+
+Due to the random initialization, the topic modelling algorithm is not entirely deterministic.
```
Comment (angel-daza) [inline episodes/02-word_representations.md]: Is there an easy way to explain why is the initialization random?

File: episodes/02-word_representations.md [hunk 11]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+Before the first training run, it downloads the language model; this and the training itself will take a few minutes.
+
+```python
+from bertopic import BERTopic
+
+# TODO: determine good number based on full data
+N_SENTENCES = 100000
+
+topic_model = BERTopic(nr_topics=100)
+topics, probs = topic_model.fit_transform(sentences[:N_SENTENCES])
```
Comment (angel-daza) [inline episodes/02-word_representations.md]: This took almost 5 minutes running in my laptop, hope it is ok for weaker machines

File: episodes/02-word_representations.md [hunk 12]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+Due to the random initialization, the topic modelling algorithm is not entirely deterministic.
+Therefore, the exact ordering and representations of the topics can be slightly different for each run.
+
+For each topic, you see:
+
+1. `Count`: the number of topics in which this topic is prevalent
+2. `Name`: by default the topic index plus the four most representative words.
+3. `Representation`: a list of the topic's most representative words
+4. `Representative_Docs`: the topic's three most representative documents from the training data.
```
Comment (angel-daza) [inline episodes/02-word_representations.md:699]: do you know how are the documents named? Is there a way to "trace them back" to the source?

File: episodes/02-word_representations.md (hunk 11 above)
Comment (angel-daza) [inline episodes/02-word_representations.md]: so, basically here document == sentence. Perhaps we can clarify that. (would be nicer to have paragraphs but of course I get it is too much distraction to define yet one more preprocessing pipeline)

File: episodes/02-word_representations.md [hunk 13]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+
+
+::: challenge
+
+### Explore Topics and Modelling Parameters
+
+1. Pick three topics and assign a concise, descriptive label to it
+2. Change the number and minimum size of topics (`BERTopic(nr_topics=..., min_topic_size=...)`) and observe how topics change
+3. Think of a way to evaluate the topic model systematically.
```
Comment (angel-daza) [inline episodes/02-word_representations.md:757]: I don't know if this last one will take up too much time (precisely because it is a very interesting question)

File: episodes/02-word_representations.md [hunk 14]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+Each of them is variable and impacts the eventual outcomes.
+The package provides reasonable defaults that typically provide a good starting point.
+Understanding the components of the algorithm allows you to adapt it to specific datasets or research questions.
+
+#### 1. Document Embeddings
+
+In the first step, we need a vector representation of each document that reflects their content.
+BERTopic makes use of existing models for that purpose so that the specifics depend on the chosen model.
+By default, it uses a [Sentence Transformers](https://sbert.net/) model called [`all-MiniLM-L6-v2`](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2), a model that has been trained to generate semantic representations for English texts with a maximum length of 256 word pieces; where each token comprises one or multiple pieces.
+This model outputs vectors with 384 dimensions.
```
Comment (angel-daza) [inline episodes/02-word_representations.md]: This is just after Word2Vec, so it is worth to clarify that here each vector is not representing individual words but a single vector represents the whole sequence of words belonging to that document. We have to be smart about this since we will introduce contextualized models until the next episode

File: episodes/02-word_representations.md [hunk 15]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+The distance between document-vectors in the vector space reflect the semantic similarity of the respective texts.
+
+Alternative Sentence Transformer models that are available on the [HuggingFace](https://huggingface.co/) platform can be used through the `embedding_model` parameter in the `BERTopic` object initialization.
```
Comment (angel-daza) [inline episodes/02-word_representations.md:793]: also "HuggingFace" is officially introduced later

File: episodes/02-word_representations.md [hunk 16]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+By choosing a different language model you can adapt the BERTopic pipeline when other languages than English, longer texts, faster processing, higher accuracy, or different modelling algorithms are needed.
+
+For corpora in which multiple languages occur, or languages for which no specific model is available, the `paraphrase-multilingual-MiniLM-L12-v2` model is recommended.
+It is loaded automatically, if `language='multilingual'` is passed to the `BERTopic` object:
+
+```python
+topic_model = BERTopic(language='multilingual')
+```
+
+Any other model available on the HuggingFace platform can be loaded via `embedding_model` -- overriding the `language` argument:
```
Comment (angel-daza) [inline episodes/02-word_representations.md:806]: maybe this is "Advanced material" to revisit later? Since HF, etc is not yet mentioned at all. Its very nice to mention the multilingual one though

File: episodes/02-word_representations.md [hunk 17]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+
+#### 2. Reduce dimensionality
+
+The document embeddings computed in the previous step are quite large.
+To make the subsequent computations less costly regarding time and resources, they are projected to a lower dimensionality.
+
+The [UMAP](https://umap-learn.readthedocs.io/en/latest/) algorithm has been developed to project vectors to a lower dimensionality while largely preserving their mutual distance -- this is the crucial property used in the next following steps.
+For instance, the 384-dimensional vectors generated by the `all-MiniLM-L6-v2` model in the first step will be represented with five dimensions per documents.
+
+Again, BERTopic offers using alternative algorithms like PCA (_Principal Component Analysis_), but UMAP provides the state-of-the-art algorithm for this specific task.
```
Comment (angel-daza) [inline episodes/02-word_representations.md:830]: I think this dimensionality part should also be an advanced material, especially manipulating the code? Explaining what UMAP is and giving a pointer to other types of dim reduction might be already complicated enough

File: episodes/02-word_representations.md [hunk 18]
```diff
@@ -650,6 +650,338 @@ Is there a strict minimum? Not really. It’s important to keep in mind that ``v
...
+Neverthless, it allows for the inherent ambiguity of words as they can be relevant for different topics.
+The word `bank`, for instance, could have a high weight in a money-related topic as well as in a river-related topic if it occurs frequently enough in those clusters, but rarely in others.
+
+At the same time, words that occur universally get a low score for all clusters for not being topic-specific.
+Nevertheless, the calculation can suffer under extreme frequencies of words like `the` and `a` in English.
+Therefore, BERTopic removes such stop words by default through a pre-defined list of English stop words.
+
+In our case, however, no action is needed here because stop words have already been removed when the data was preprocessed above.
+
+### Model Tuning
```
Comment (angel-daza) [inline episodes/02-word_representations.md:890]: In general the "Step by Step" section is well written and is a nice guide for customization, but I am worried about how much sense it makes to people before talking at all about Transformers; the leap from word2vec to customizing algorithms at this level is too big in my opinion. We should discuss this more later

**Review by angel-daza — COMMENTED** [review]
In general looks nice and very complete! My biggest worry is the "step by step" section which might be going too deep into technical decisions and might confuse and take too much time, especially for the end of a session

File: episodes/02-word_representations.md (hunk 1 above)
Comment (carschno) [reply episodes/02-word_representations.md]: I was actually wondering what to use here. In this case, I chose `text` because `document` seems to imply a longer, complete unit, e.g. a book, whereas we use sentences here. On the other hand, `text` might be similarly misleading. 🤔

File: episodes/02-word_representations.md (hunk 4 above)
Comment (carschno) [reply episodes/02-word_representations.md]: Good catch, removing that. If we want to explain self-supervised learning, it is probably better located in the language modelling section.

File: episodes/02-word_representations.md (hunk 16 above)
Comment (carschno) [reply episodes/02-word_representations.md:806]: From our discussion: I will leave this part as it is for now, but keep it optional during the 2-day workshop, depending on time and audience interest.

File: episodes/02-word_representations.md (hunk 8 above)
Comment (carschno) [reply episodes/02-word_representations.md:657]: It is, so that learners know what output to expect.

File: episodes/02-word_representations.md (hunk 12 above)
Comment (carschno) [reply episodes/02-word_representations.md:699]: It is just the beginning of each document (or the entire doc if it is short enough).

File: episodes/02-word_representations.md (hunk 13 above)
Comment (carschno) [reply episodes/02-word_representations.md:757]: I have added a few approaches, but it should mostly illustrate how difficult (and important) this is.
